example/src/main/scala/tamer/example/RestCustomAuth.scala

```diff
@@ -0,0 +1,65 @@
+package tamer.example
+
+import sttp.client3.httpclient.zio.{HttpClientZioBackend, SttpClient, send}
+import sttp.client3.{UriContext, basicRequest}
+import tamer.config.{Config, KafkaConfig}
+import tamer.rest.TamerRestJob.Offset
+import tamer.rest.{Authentication, DecodedPage, LocalSecretCache, SttpRequest, TamerRestJob}
+import tamer.{AvroCodec, TamerError}
+import zio.{ExitCode, Has, Layer, RIO, Ref, Task, URIO, ZEnv, ZIO, ZLayer}
+
+object RestCustomAuth extends zio.App {
+  val httpClientLayer: RLayer[ZEnv, SttpClient] =
+    HttpClientZioBackend.layer()
+  val kafkaConfigLayer: Layer[TamerError, KafkaConfig] = Config.live
+  val fullLayer: RLayer[ZEnv, SttpClient with KafkaConfig with Has[Ref[Option[String]]]] =
+    httpClientLayer ++ kafkaConfigLayer ++ LocalSecretCache.live
+
+  case class MyData(i: Int)
+
+  object MyData {
+    implicit val codec = AvroCodec.codec[MyData]
+  }
+
+  case class MyKey(i: Int)
+
+  object MyKey {
+    implicit val codec = AvroCodec.codec[MyKey]
+  }
+
+  val pageDecoder: String => Task[DecodedPage[MyData, Offset]] = DecodedPage.fromString { pageBody =>
+    val dataRegex = """.*"data":"(-?[\d]+).*""".r
+    pageBody match {
+      case dataRegex(data) => Task(List(MyData(data.toInt)))
+      case _               => Task.fail(new RuntimeException(s"Could not parse pageBody: $pageBody"))
+    }
+  }
+
+  val authentication: Authentication[SttpClient] = new Authentication[SttpClient] {
+
+    override def addAuthentication(request: SttpRequest, bearerToken: Option[String]): SttpRequest = request.auth.bearer(bearerToken.getOrElse(""))
+
+    override def setSecret(secretRef: Ref[Option[String]]): ZIO[SttpClient, TamerError, Unit] = {
+      val fetchToken = send(basicRequest.get(uri"http://localhost:9095/auth").auth.basic("user", "pass"))
+        .flatMap(_.body match {
+          case Left(error)  => ZIO.fail(TamerError(error))
+          case Right(token) => ZIO.succeed(token)
+        })
```

Comment:
```suggestion
        .flatMap(response => ZIO.fromEither(response.body))
```

Reply:
This saves two lines but then I'll have to convert the error into something else (the LUB between String and Throwable is serializable) so at the end of the day there is not a lot of difference.
This should be taken care of if we ever decide to uniform all the errors in Tamer, there are a few things out of place (Strings in the rest module) and a few TamerError invading the user land.

Reply:
Sorry, hadn't realised that in `Left(error)`, `error: String` 🤦